File: scripts/generate_swagger_docs.py
#!/usr/bin/env python3
"""
Generate Swagger/OpenAPI documentation from Go source code.

This script parses Go source files to extract API endpoint information
and generates or updates the Swagger/OpenAPI specification.
"""
import os
import re
import json
import argparse
from pathlib import Path
from typing import Dict, List, Optional, Any, Tuple

# Configuration
PROJECT_ROOT = Path(__file__).parent.parent
SWAGGER_FILE = PROJECT_ROOT / "docs" / "swagger" / "swagger.json"
HANDLERS_DIR = PROJECT_ROOT / "internal" / "handlers"
MODELS_DIR = PROJECT_ROOT / "internal" / "constants" / "dto"

# Regex patterns to extract API information
ROUTE_PATTERN = re.compile(r'router\.(?:Get|Post|Put|Delete|Patch|Options|Head|Connect|Trace)\([^,]+,\s*"([^"]+)"'
                          r'|r\.(?:Get|Post|Put|Delete|Patch|Options|Head|Connect|Trace)\([^,]+,\s*"([^"]+)"'
                          r'|chi\.Router\(\)\.(?:Get|Post|Put|Delete|Patch|Options|Head|Connect|Trace)\([^,]+,\s*"([^"]+)"'
                          r'|r\.With\([^)]+\)\.(?:Get|Post|Put|Delete|Patch|Options|Head|Connect|Trace)\([^,]+,\s*"([^"]+)"')

HANDLER_PATTERN = re.compile(r'func\s+([a-zA-Z0-9_]+)\s*\(')
COMMENT_PATTERN = re.compile(r'//\s*@(\w+)\s*(.*)')

# Swagger template
SWAGGER_TEMPLATE = {
    "openapi": "3.0.0",
    "info": {
        "title": "CPS Action API",
        "description": "API documentation for the CPS Action module of CBE Super App",
        "version": "1.0.0",
        "contact": {
            "name": "API Support",
            "email": "contact@eaglelionsystems.com"
        }
    },
    "servers": [
        {
            "url": "http://localhost:8080/api/v1/cbesuperapp/cps_action",
            "description": "Local Development Server"
        }
    ],
    "paths": {},
    "components": {
        "securitySchemes": {
            "BearerAuth": {
                "type": "http",
                "scheme": "bearer",
                "bearerFormat": "JWT"
            }
        },
        "schemas": {},
        "responses": {}
    },
    "security": [{"BearerAuth": []}]
}

class SwaggerGenerator:
    def __init__(self):
        self.swagger = SWAGGER_TEMPLATE.copy()
        self.current_path = ""
        self.current_method = ""
        self.current_component = ""
        
    def load_existing_swagger(self) -> bool:
        """Load existing Swagger file if it exists."""
        if SWAGGER_FILE.exists():
            try:
                with open(SWAGGER_FILE, 'r', encoding='utf-8') as f:
                    self.swagger = json.load(f)
                return True
            except json.JSONDecodeError:
                print(f"Warning: Could not parse {SWAGGER_FILE}. Creating a new one.")
        return False
    
    def save_swagger(self):
        """Save the Swagger specification to a file."""
        # Create directory if it doesn't exist
        SWAGGER_FILE.parent.mkdir(parents=True, exist_ok=True)
        
        # Write the Swagger file with pretty-printing
        with open(SWAGGER_FILE, 'w', encoding='utf-8') as f:
            json.dump(self.swagger, f, indent=2, ensure_ascii=False)
        
        print(f"Swagger documentation generated at {SWAGGER_FILE}")
    
    def process_go_files(self):
        """Process all Go files to extract API information."""
        # Process handler files
        for root, _, files in os.walk(HANDLERS_DIR):
            for file in files:
                if file.endswith('.go') and not file.endswith('_test.go'):
                    self.process_go_file(Path(root) / file)
        
        # Process model files
        for root, _, files in os.walk(MODELS_DIR):
            for file in files:
                if file.endswith('.go') and not file.endswith('_test.go'):
                    self.process_model_file(Path(root) / file)
    
    def process_go_file(self, file_path: Path):
        """Process a Go file to extract API endpoint information."""
        print(f"Processing {file_path}...")
        
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # Extract route information
        for match in ROUTE_PATTERN.finditer(content):
            path = match.group(1) or match.group(3) or match.group(5) or match.group(7)
            method = match.group(0).split('.')[-1].split('(')[0].lower()
            
            # Initialize path if it doesn't exist
            if path not in self.swagger["paths"]:
                self.swagger["paths"][path] = {}
            
            # Add method to path
            self.swagger["paths"][path][method] = {
                "tags": [file_path.stem],
                "summary": f"{method.upper()} {path}",
                "responses": {
                    "200": {
                        "description": "Success",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/StandardResponse"
                                }
                            }
                        }
                    },
                    "default": {
                        "$ref": "#/components/responses/InternalServerError"
                    }
                },
                "security": [{"BearerAuth": []}]
            }
    
    def process_model_file(self, file_path: Path):
        """Process a Go file to extract model information."""
        print(f"Processing model {file_path}...")
        
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # This is a simplified example - in a real implementation, you would parse the Go structs
        # and convert them to JSON Schema
        
        # For now, we'll just add a placeholder for the model
        model_name = file_path.stem
        if model_name not in self.swagger["components"]["schemas"]:
            self.swagger["components"]["schemas"][model_name] = {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "string",
                        "description": f"Unique identifier for the {model_name}"
                    }
                }
            }

def main():
    parser = argparse.ArgumentParser(description='Generate Swagger/OpenAPI documentation from Go source code.')
    parser.add_argument('--output', '-o', default=str(SWAGGER_FILE),
                      help=f'Output file path (default: {SWAGGER_FILE})')
    args = parser.parse_args()
    
    # Initialize the generator
    generator = SwaggerGenerator()
    
    # Load existing Swagger file if it exists
    generator.load_existing_swagger()
    
    # Process Go files to extract API information
    generator.process_go_files()
    
    # Save the updated Swagger documentation
    generator.save_swagger()
    
    print("Swagger documentation generation complete!")

if __name__ == "__main__":
    main()

File: scripts/generate_swagger_docs.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

type SwaggerDoc struct {
	OpenAPI    string                    `json:"openapi"`
	Info       Info                      `json:"info"`
	Servers    []Server                  `json:"servers"`
	Paths      map[string]map[string]any `json:"paths"`
	Components Components                `json:"components"`
}

type Info struct {
	Title       string `json:"title"`
	Description string `json:"description"`
	Version     string `json:"version"`
	Contact     struct {
		Name  string `json:"name,omitempty"`
		Email string `json:
```go
"email,omitempty"`
	} `json:"contact,omitempty"`
}

type Server struct {
	URL         string `json:"url"`
	Description string `json:"description"`
}

type Components struct {
	SecuritySchemes map[string]SecurityScheme `json:"securitySchemes,omitempty"`
	Schemas         map[string]any            `json:"schemas,omitempty"`
}

type SecurityScheme struct {
	Type         string `json:"type"`
	Scheme       string `json:"scheme,omitempty"`
	BearerFormat string `json:"bearerFormat,omitempty"`
}

func main() {
	handlersDir := flag.String("handlers", "internal/handlers/rest/http", "Path to the handlers directory")
	outputDir := flag.String("output", "docs/swagger", "Output directory for Swagger files")
	flag.Parse()

	// Create output directory if it doesn't exist
	if err := os.MkdirAll(*outputDir, 0755); err != nil {
		log.Fatalf("Failed to create output directory: %v", err)
	}

	// Process each module in the handlers directory
	modules, err := os.ReadDir(*handlersDir)
	if err != nil {
		log.Fatalf("Failed to read handlers directory: %v", err)
	}

	for _, module := range modules {
		if !module.IsDir() {
			continue
		}

		modulePath := filepath.Join(*handlersDir, module.Name())
		doc := createBaseSwaggerDoc(module.Name())
		processModule(modulePath, doc, module.Name())

		// Write the Swagger JSON file
		outputFile := filepath.Join(*outputDir, module.Name()+".json")
		jsonData, err := json.MarshalIndent(doc, "", "  ")
		if err != nil {
			log.Printf("Failed to marshal Swagger JSON for %s: %v", module.Name(), err)
			continue
		}

		if err := os.WriteFile(outputFile, jsonData, 0644); err != nil {
			log.Printf("Failed to write Swagger file for %s: %v", module.Name(), err)
			continue
		}

		log.Printf("Generated Swagger documentation for %s at %s", module.Name(), outputFile)
	}
}

func createBaseSwaggerDoc(moduleName string) *SwaggerDoc {
	doc := &SwaggerDoc{
		OpenAPI: "3.0.0",
		Info: Info{
			Title:       fmt.Sprintf("%s API", strings.Title(moduleName)),
			Description: fmt.Sprintf("API documentation for the %s module", moduleName),
			Version:     "1.0.0",
			Contact: struct {
				Name  string `json:"name,omitempty"`
				Email string `json:"email,omitempty"`
			}{
				Name:  "API Support",
				Email: "contact@eaglelionsystems.com",
			},
		},
		Servers: []Server{
			{
				URL:         "http://localhost:8080/api/v1/cbesuperapp/cps_action",
				Description: "Local Development Server",
			},
		},
		Paths: make(map[string]map[string]any),
		Components: Components{
			SecuritySchemes: map[string]SecurityScheme{
				"BearerAuth": {
					Type:         "http",
					Scheme:       "bearer",
					BearerFormat: "JWT",
				},
			},
			Schemas: make(map[string]any),
		},
	}

	return doc
}

func processModule(modulePath string, doc *SwaggerDoc, moduleName string) {
	// Walk through all Go files in the module directory
	err := filepath.Walk(modulePath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Skip non-Go files and test files
		if info.IsDir() || !strings.HasSuffix(path, ".go") || strings.HasSuffix(path, "_test.go") {
			return nil
		}

		// Parse the Go file
		fset := token.NewFileSet()
		node, err := parser.ParseFile(fset, path, nil, parser.ParseComments)
		if err != nil {
			log.Printf("Failed to parse file %s: %v", path, err)
			return nil
		}

		// Process each function in the file
		ast.Inspect(node, func(n ast.Node) bool {
			// Look for function declarations
			fn, ok := n.(*ast.FuncDecl)
			if !ok || fn.Doc == nil {
				return true
			}

			// Extract API documentation from comments
			for _, comment := range fn.Doc.List {
				text := comment.Text
				if strings.HasPrefix(text, "// @") {
					// Process Swagger annotations
					processSwaggerAnnotation(text, doc, moduleName)
				}
			}

			return true
		})

		return nil
	})

	if err != nil {
		log.Printf("Error processing module %s: %v", moduleName, err)
	}
}

func processSwaggerAnnotation(annotation string, doc *SwaggerDoc, moduleName string) {
	// This is a simplified example - in a real implementation, you would parse
	// the Swagger annotations and update the Swagger document accordingly
	// For now, we'll just log the annotation
	log.Printf("Found Swagger annotation in %s: %s", moduleName, annotation)
}
